designs/0032-cmdstan-output-formats.md

```diff
@@ -0,0 +1,230 @@
+- Feature Name: cmdstan-output-formats
+- Start Date: 2022-01-11
+- RFC PR:
+- Stan Issue:
+
+# Summary
+[summary]: #summary
+
+Provide alternatives to the [Stan CSV file format](https://mc-stan.org/docs/cmdstan-guide/stan_csv.html)
+for outputs of the Stan platform services methods supported by CmdStan.
+
+# Motivation
+[motivation]: #motivation
+
+The [Stan CSV file format](https://mc-stan.org/docs/cmdstan-guide/stan_csv.html) is limited and limiting for several reasons.
+
+The CSV format itself is not fully standardized.
+The original intent of the format is to represent a table's worth of data in plain text.
+Each row of the file contains one row of table data, and all rows have the same number of fields.
+Over time, conventions have been adopted, notably use of row headers for column labels and comment rows.
+
+The Stan CSV file, however, contains far more information than a single table's worth of data.
+To cram all the information from a CmdStan run we 
+abuse the CSV format by using blocks of comment lines to record additional information.
+The current Stan API requires the information in a comment line be assembled and flattened into a text string.
+Even when we can use a fast CSV parser to extract the CSV data, we still need to build ad-hoc parsers for the comment blocks
+to re-assemble the flattened information.
+
+The sample methods produce the largest amount of output data and these are the most difficult to process quickly and efficiently.
+The first part of a CmdStan CSV file always a series of comment rows which contain configuration and version information,
+and this comment header continues to accumulate new bits of information about program version and compilation options.
+The NUTS-HMC methods output a series of comment lines which contain the stepsize and metric.
+The sample methods also record timing information at the end.
+
+In order to monitor the progress of CmdStan's inference algorithms,
+the outputs from CmdStan must be available to downstream readers on a streaming basis.
+To facilitate monitoring different aspects of the algorithm, a more granular set of outputs,
+instead of a single CSV file would make this task easier.
+Another consideration is the use of multi-chain methods; currently multi-chain NUTS-HMC and in the future,
+multi-chain Pathfinder.  When running any kind of ensemble method, we need to record the process/chain_id
+and make this available 
+
+Therefore we need to re-implement the set of output methods used by CmdStan and the core Stan services
+to capture more information in a more structured way.
+At the same time, we need to continue to support the current Stan CSV format in order to support existing downstream processing packages.
```

Comment:
I think you should move this up front as a top-level goal of backward compatibility.

Reply:
moved.